src/flow/src/batching_mode/engine.rs

```diff
@@ -0,0 +1,1151 @@
+            query_ctx.clone(),
+        )
+        .await?;
+
+        let phy_expr = time_window_expr
+            .map(|expr| {
+                TimeWindowExpr::from_expr(
+                    &expr,
+                    &column_name,
+                    &df_schema,
+                    &self.query_engine.engine_state().session_state(),
+                )
+            })
+            .transpose()?;
+
+        info!("Flow id={}, found time window expr={:?}", flow_id, phy_expr);
+
+        let task = BatchingTask::new(
+            flow_id,
+            &sql,
+            plan,
+            phy_expr,
+            expire_after,
+            sink_table_name,
+            source_table_names,
+            query_ctx,
+            self.table_meta.clone(),
+            rx,
+        );
+
+        let task_inner = task.clone();
+        let engine = self.query_engine.clone();
+        let frontend = self.frontend_client.clone();
+
+        task.check_execute(&engine, &frontend).await?;
+
+        // TODO(discord9): also save handle & use time wheel or what for better
+        let _handle = common_runtime::spawn_global(async move {
+            match task_inner.start_executing_loop(engine, frontend).await {
+                Ok(()) => info!("Flow {} shutdown", task_inner.flow_id),
+                Err(err) => common_telemetry::error!(
+                    "Flow {} encounter unrecoverable error: {err:?}",
+                    task_inner.flow_id
+                ),
+            }
+        });
+
+        // only replace here not earlier because we want the old one intact if something went wrong before this line
+        let replaced_old_task_opt = self.tasks.write().await.insert(flow_id, task);
+        drop(replaced_old_task_opt);
+
+        self.shutdown_txs.write().await.insert(flow_id, tx);
+
+        Ok(Some(flow_id))
+    }
+
+    pub async fn remove_flow(&self, flow_id: FlowId) -> Result<(), Error> {
+        if self.tasks.write().await.remove(&flow_id).is_none() {
+            warn!("Flow {flow_id} not found in tasks")
+        }
+        let Some(tx) = self.shutdown_txs.write().await.remove(&flow_id) else {
+            UnexpectedSnafu {
+                reason: format!("Can't found shutdown tx for flow {flow_id}"),
+            }
+            .fail()?
+        };
+        if tx.send(()).is_err() {
+            warn!("Fail to shutdown flow {flow_id} due to receiver already dropped, maybe flow {flow_id} is already dropped?")
+        }
+        Ok(())
+    }
+
+    pub async fn flush_flow(&self, flow_id: FlowId) -> Result<(), Error> {
+        let task = self.tasks.read().await.get(&flow_id).cloned();
+        let task = task.with_context(|| UnexpectedSnafu {
+            reason: format!("Can't found task for flow {flow_id}"),
+        })?;
+
+        task.gen_exec_once(&self.query_engine, &self.frontend_client)
+            .await?;
+        Ok(())
+    }
+
+    pub async fn flow_exist(&self, flow_id: FlowId) -> bool {
```

Comment:
Add documents to all these pub functions. And I wonder if we can add the flow engine and task trait?